pkg/controller/seed-controller-manager/initial-application-installation-controller/controller.go

```diff
@@ -0,0 +1,231 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package initialapplicationinstallationcontroller
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/go-openapi/errors"
+	"go.uber.org/zap"
+
+	v1 "k8c.io/kubermatic/v2/pkg/api/v1"
+	appkubermaticv1 "k8c.io/kubermatic/v2/pkg/apis/apps.kubermatic/v1"
+	kubermaticv1 "k8c.io/kubermatic/v2/pkg/apis/kubermatic/v1"
+	kubermaticv1helper "k8c.io/kubermatic/v2/pkg/apis/kubermatic/v1/helper"
+	clusterclient "k8c.io/kubermatic/v2/pkg/cluster/client"
+	predicateutil "k8c.io/kubermatic/v2/pkg/controller/util/predicate"
+	"k8c.io/kubermatic/v2/pkg/provider"
+	"k8c.io/kubermatic/v2/pkg/version/kubermatic"
+
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/record"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	ControllerName = "kkp-initial-application-installation-controller"
+)
+
+// UserClusterClientProvider provides functionality to get a user cluster client.
+type UserClusterClientProvider interface {
+	GetClient(ctx context.Context, c *kubermaticv1.Cluster, options ...clusterclient.ConfigOption) (ctrlruntimeclient.Client, error)
+}
+
+type Reconciler struct {
+	ctrlruntimeclient.Client
+
+	workerName                    string
+	recorder                      record.EventRecorder
+	seedGetter                    provider.SeedGetter
+	userClusterConnectionProvider UserClusterClientProvider
+	log                           *zap.SugaredLogger
+	versions                      kubermatic.Versions
+}
+
+func Add(ctx context.Context, mgr manager.Manager, numWorkers int, workerName string, seedGetter provider.SeedGetter, userClusterConnectionProvider UserClusterClientProvider, log *zap.SugaredLogger, versions kubermatic.Versions) error {
+	reconciler := &Reconciler{
+		Client: mgr.GetClient(),
+
+		workerName:                    workerName,
+		recorder:                      mgr.GetEventRecorderFor(ControllerName),
+		seedGetter:                    seedGetter,
+		userClusterConnectionProvider: userClusterConnectionProvider,
+		log:                           log,
+		versions:                      versions,
+	}
+
+	c, err := controller.New(ControllerName, mgr, controller.Options{
+		Reconciler:              reconciler,
+		MaxConcurrentReconciles: numWorkers,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create controller: %w", err)
+	}
+
+	predicate := predicateutil.MultiFactory(predicateutil.TrueFilter, func(o ctrlruntimeclient.Object) bool {
+		_, exists := o.GetAnnotations()[v1.InitialApplicationInstallationsRequestAnnotation]
+		return exists
+	}, predicateutil.TrueFilter)
```

Comment:
I am a bit confused by this predicate. If I understand it correctly, it automatically lets through any Create and Delete events, but will check the annotation for Update events. However in my opinion, we should also check for the annotation for Create and Delete events, so we do not trigger an unnecessary reconciliation on cluster objects that are missing the annotation, right?

Reply:
It checks for all the event types i.e. create, update and delete.